### PR TITLE
Non-record: Faithful Conditional Memory

### PR DESCRIPTION
**Pull request by wisebreadloaf**
## Summary

Adds a non-record 16MB submission for a faithful standalone conditional-memory architecture.

This submission explores:
- tokenizer-normalized compressed lookup identities
- multi-hash n-gram memory tables
- contextual key/value projections
- internal-layer memory injection

## Result
- 8xH100 result:
  - `final_int8_zlib_roundtrip_exact val_bpb: 1.61095535`
- Artifact size:
  - `2,773,498` bytes

## Why submit this
This method showed strong standalone local signs of life before cloud evaluation, but did not transfer competitively enough to challenge the dense frontier. It is submitted as a non-record methodological contribution and negative/partial result for future work.

## Included
- `README.md`
- `submission.json`
- `train_gpt.py`
- cloud log

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Non-record: Faithful Conditional Memory

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## Analysis PR #1490 ("FaithfulConditionalMemory") introduces a `MemoryGPT` architecture with n-gram engram memory (`EngramMemory`, `NgramHashMappingTorch`) layered into a standard causal transformer. The submission was checked against all three illegal patterns. ### N-gram family bug (target XOR'd into hash lookup key) — NOT PRESENT The `NgramHashMappingTorch.forward()` hashes only **input** token context (left-shifted windows from `input_ids`). Lines 873–895: ```python def forward(self, input_ids: Tensor, layer_id: int) -> Tensor: x = self.lookup_table[input_ids.long()] # context only for k in range(self.max_ngram_size): shifts.append(torch.cat((pad, x[:, :-k]), dim=1)) # left shifts for n in range(2, self.max_ngram_size + 1): mix = shifts[0] * multipliers[0] for k in range(1, n): mix = torch.bitwise_xor(mix, shifts[k] * multipliers[k]) # ctx XOR ctx only all_hashes.append(torch.remainder(mix, head_sizes[head_idx])) ``` No target token IDs are mixed into the hash key at any point. The XOR is purely context-token × context-token, identical in spirit to the legal `BigramHash` pattern. The `EngramMemory.forward()` (lines 973–986) similarly receives only `hidden_states` and `input_ids` — never `target_ids`. ### Pre-Quant TTT (multi-epoch AdamW on val_tokens without score-first) — NOT PRESENT There is no optimizer update loop over `val_tokens`. The two val eval functions (`eval_val`, `eval_val_sliding`) run entirely under `torch.inference_mode()` (lines 303, 1159). No `.backward()` or `.step()` call is ever issued against validation data anywhere in the file. ### Scored-region SLOT (mask + optimize + score same region) — NOT PRESENT No masking + per-region optimize + score pattern exists. Training is a standard single-pass loop (lines 1540–1606) on fresh training batches; final eval is on a separate held-out `val_tokens` set loaded from `val_files`. ###...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*

**Comment by wisebreadloaf** [comment]
Thanks for re-checking and verifying. @MatoTeziTanka
